File: docs/examples/android/app/src/main/java/com/cossacklabs/themis/android/example/SecSessionExampleClient.java
package com.cossacklabs.themis.android.example;

import android.util.Base64;
import android.util.Log;

import com.cossacklabs.themis.ISessionCallbacks;
import com.cossacklabs.themis.NullArgumentException;
import com.cossacklabs.themis.PrivateKey;
import com.cossacklabs.themis.PublicKey;
import com.cossacklabs.themis.SecureSession;
import com.cossacklabs.themis.SecureSessionException;

import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.concurrent.Executors;

// ---------------------- IMPORTANT SETUP ---------------------------------------

// User id, Server id and Server Public Key should be copied from the Server Setup Page
// https://themis.cossacklabs.com/interactive-simulator/setup/
// Server public key ("server key"),

class SecSessionExampleClient {

    private final static String LOG_TAG = "SSC";

    private final static String CLIENT_ID = "ULzsfYGzScUisNV";
    private final static String SERVER_ID = "mLAResUZAQcQXXl";
    private final static String CLIENT_PRIVATE_KEY = "UkVDMgAAAC1EvnquAPUxxwJsoJxoMAkEF7c06Fo7dVwnWPnmNX5afyjEEGmG";
    private final static String CLIENT_PUBLIC_KEY = "VUVDMgAAAC18urRTA1H1hts93vlLXX59OuyVnY1tGFxl/F3PkhDtzrdQETMi";
    private final static String SERVER_PUBLIC_KEY = "VUVDMgAAAC0boM1EAvAkoWsfqbMvugv/YzzMPC6AeKT/t5gtCb3xyPpEJJB/";

    private final static String MESSAGE = "msg msg msg :)";
    private final static String URL = "https://themis.cossacklabs.com/api/ULzsfYGzScUisNV/";

    private final static Charset CHARSET = StandardCharsets.UTF_8;

    private final HttpClient client = new HttpClient(Executors.newSingleThreadExecutor());

    private void startSession(final SecureSession ss, String base64MessageToSend) {
        client.sendMessageAsync(URL, base64MessageToSend, CHARSET, new HttpCallback() {
            @Override
            public void onSuccess(byte[] response) {
                try {
                    byte[] unwrappedData = ss.unwrap(response).getData();
                    String unwrappedResult = new String(unwrappedData, StandardCharsets.UTF_8);
                    System.out.println(unwrappedResult);

                    Log.d(LOG_TAG, "session establish in progress = " + unwrappedResult);
                    String messageToSend2 = Base64.encodeToString(unwrappedData, Base64.NO_WRAP);

                    if (!ss.isEstablished()) {
                        startSession(ss, messageToSend2);
                    } else {
                        sendRealMessage(ss);
                    }

                } catch (Exception e) {
                    onFail(e);
                }
            }

            @Override
            public void onFail(Exception ex) {
                Log.e(LOG_TAG, "startSession", ex);
            }
        });
    }

    private void sendRealMessage(final SecureSession ss) throws SecureSessionException, NullArgumentException {
        byte[] wrappedMessage = ss.wrap(MESSAGE.getBytes(CHARSET));
        String base64Message = Base64.encodeToString(wrappedMessage, Base64.NO_WRAP);

        client.sendMessageAsync(URL, base64Message, CHARSET, new HttpCallback() {
            @Override
            public void onSuccess(byte[] response) {
                try {
                    byte[] unwrappedData = ss.unwrap(response).getData();
                    String unwrappedResult = new String(unwrappedData, StandardCharsets.UTF_8);
                    System.out.println(unwrappedResult);

                    Log.d(LOG_TAG, "message received = " + unwrappedResult);
                } catch (Exception e) {
                    onFail(e);
                }
            }

            @Override
            public void onFail(Exception ex) {
                Log.e(LOG_TAG, "sendRealMessage", ex);
            }
        });
    }

    void testSSessionCIClient() throws SecureSessionException {
        ISessionCallbacks callbacks = new ISessionCallbacks() {

            @Override
            public PublicKey getPublicKeyForId(SecureSession session, byte[] id) {
                byte[] serverId = SERVER_ID.getBytes(CHARSET);
                if (Arrays.equals(id, serverId)) {
                    PublicKey publicKey = new PublicKey(Base64.decode(SERVER_PUBLIC_KEY.getBytes(CHARSET), Base64.NO_WRAP));
                    Log.d(LOG_TAG, "publicKey1 = " + Arrays.toString(publicKey.toByteArray()));
                    return publicKey;
                }
                return null;
            }

            @Override
            public void stateChanged(SecureSession session) {
            }
        };

        PrivateKey privateKey = new PrivateKey(Base64.decode(CLIENT_PRIVATE_KEY.getBytes(CHARSET), Base64.NO_WRAP));
        Log.d(LOG_TAG, "privateKey1 = " + Arrays.toString(privateKey.toByteArray()));

        final SecureSession ss = new SecureSession(CLIENT_ID.getBytes(CHARSET), privateKey, callbacks);

        final String messageToSend = Base64.encodeToString(ss.generateConnectRequest(), Base64.NO_WRAP);
        startSession(ss, messageToSend);
    }
}

File: docs/examples/android/app/src/main/java/com/cossacklabs/themis/android/example/SecSessionExampleClient.kt
package com.cossacklabs.themis.android.example

import android.util.Base64
import android.util.Log
import com.cossacklabs.themis.ISessionCallbacks
import com.cossacklabs.themis.PrivateKey
import com.cossacklabs.themis.PublicKey
import com.cossacklabs.themis.SecureSession
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.concurrent.Executors

// ---------------------- IMPORTANT SETUP ---------------------------------------
// User id, Server id and Server Public Key should be copied from the Server Setup Page
// https://themis.cossacklabs.com/interactive-simulator/setup/
// Server public key ("server key"),
internal class SecSessionExampleClient {
    private val client = HttpClient(Executors.newSingleThreadExecutor())
    private fun startSession(ss: SecureSession, base64MessageToSend: String) {
        client.sendMessageAsync(URL, base64MessageToSend, CHARSET, object : HttpCallback {
            override fun onSuccess(response: ByteArray?) {
                try {
                    val unwrappedData = ss.unwrap(response).data
                    val unwrappedResult = String(unwrappedData, StandardCharsets.UTF_8)
                    println(unwrappedResult)
                    Log.d(LOG_TAG, "session establish in progress = $unwrappedResult")
                    val messageToSend2 = Base64.encodeToString(unwrappedData, Base64.NO_WRAP)
                    if (!ss.isEstablished) {
                        startSession(ss, messageToSend2)
                    } else {
                        sendRealMessage(ss)
                    }
                } catch (e: Exception) {
                    onFail(e)
                }
            }

            override fun onFail(ex: Exception?) {
                Log.e(LOG_TAG, "startSession", ex)
            }
        })
    }

    private fun sendRealMessage(ss: SecureSession) {
        val wrappedMessage = ss.wrap(MESSAGE.toByteArray(CHARSET))
        val base64Message = Base64.encodeToString(wrappedMessage, Base64.NO_WRAP)
        client.sendMessageAsync(URL, base64Message, CHARSET, object : HttpCallback {
            override fun onSuccess(response: ByteArray?) {
                try {
                    val unwrappedData = ss.unwrap(response).data
                    val unwrappedResult = String(unwrappedData, StandardCharsets.UTF_8)
                    println(unwrappedResult)
                    Log.d(LOG_TAG, "message received = $unwrappedResult")
                } catch (e: Exception) {
                    onFail(e)
                }
            }

            override fun onFail(ex: Exception?) {
                Log.e(LOG_TAG, "sendRealMessage", ex)
            }
        })
    }

    fun testSSessionCIClient() {
        val callbacks: ISessionCallbacks = object : ISessionCallbacks {
            override fun getPublicKeyForId(session: SecureSession, id: ByteArray): PublicKey? {
                val serverId = SERVER_ID.toByteArray(CHARSET)
                if (id.contentEquals(serverId)) {
                    val publicKey = PublicKey(Base64.decode(SERVER_PUBLIC_KEY.toByteArray(CHARSET), Base64.NO_WRAP))
                    Log.d(LOG_TAG, "publicKey1 = " + Arrays.toString(publicKey.toByteArray()))
                    return publicKey
                }
                return null
            }

            override fun stateChanged(session: SecureSession) {}
        }
        val privateKey = PrivateKey(Base64.decode(CLIENT_PRIVATE_KEY.toByteArray(CHARSET), Base64.NO_WRAP))
        Log.d(LOG_TAG, "privateKey1 = " + Arrays.toString(privateKey.toByteArray()))
        val ss = SecureSession(CLIENT_ID.toByteArray(CHARSET), privateKey, callbacks)
        val messageToSend = Base64.encodeToString(ss.generateConnectRequest(), Base64.NO_WRAP)
        startSession(ss, messageToSend)
    }

    companion object {
        private const val LOG_TAG = "SSC"
        private const val CLIENT_ID = "ULzsfYGzScUisNV"
        private const val SERVER_ID = "mLAResUZAQcQXXl"
        private const val CLIENT_PRIVATE_KEY = "UkVDMgAAAC1EvnquAPUxxwJsoJxoMAkEF7c06Fo7dVwnWPnmNX5afyjEEGmG"
        private const val CLIENT_PUBLIC_KEY = "VUVDMgAAAC18urRTA1H1hts93vlLXX59OuyVnY1tGFxl/F3PkhDtzrdQETMi"
        private const val SERVER_PUBLIC_KEY = "VUVDMgAAAC0boM1EAvAkoWsfqbMvugv/YzzMPC6AeKT/t5gtCb3xyPpEJJB/"
        private const val MESSAGE = "msg msg msg :)"
        private const val URL = "https://themis.cossacklabs.com/api/ULzsfYGzScUisNV/"
        private val CHARSET = StandardCharsets.UTF_8
    }
}
